### ISSUE TITLE
Extend http interceptors in child module injector

### ISSUE THREAD

**Issue by miguel-leon** [angular/angular#36974]
# 🚀 feature request

### Relevant Package
This feature request is for @angular/common/http

### Description
Ability to extend http interceptors in child module injectors since it is fairly simple, but the api to achieve it is "private" in angular.

### Describe the solution you'd like

I've already managed to do this by providing `HttpHandler`, `HttpClient` and  `HTTP_INTERCEPTORS` manually in the lazy loaded module without importing `HttpClientModule`, since this module imports `HttpClientXsrfModule` which provides a fresh `HTTP_INTERCEPTORS` array, for the sake of reincluding `HttpXsrfInterceptor` (imho, but why again?).

If this explanation was hard to follow, here's the code
```typescript
@NgModule({
	providers: [{
		provide: HttpClient,
		useClass: HttpClient
	}, {
		provide: HttpHandler,
		useClass: ɵHttpInterceptingHandler
	}, {
		provide: HTTP_INTERCEPTORS,
		deps: [Injector],
		useFactory: injector => [
			...ARRAY_OF_CHILDREN_INTERCEPTOR,
			...injector.parent.get(HTTP_INTERCEPTORS)
		]
	}]
})
export class FeatureModule {}
```

Here the "private" angular api is the class `ɵHttpInterceptingHandler` and the `injector.parent` part.

So, I think the most practical solution would be to maybe have an alternative `HttpClientModule` that does not import `HttpClientXsrfModule` so the interceptors array is not reset.

Or similarly, have both these modules without providers and have `.forRoot(): ModuleWithProviders` options for each, of which `HttpClientModule.forRoot()` would spread `HttpClientXsrfModule.forRoot()`'s providers.

Code again, if explanation was convoluted
```typescript
class HttpClientModule {
	static forRoot(): ModuleWithProviders<HttpClientModule> {
		return {
			ngModule: HttpClientModule,
			providers: [
				// taken from source code as of version 9
				HttpClient,
				{provide: HttpHandler, useClass: HttpInterceptingHandler},
				HttpXhrBackend,
				{provide: HttpBackend, useExisting: HttpXhrBackend},
				BrowserXhr,
				{provide: XhrFactory, useExisting: BrowserXhr},

				// HttpClientXsrfModule providers
				...HttpClientXsrfModule.forRoot().providers
			]
		}
	}
}
```

### Describe alternatives you've considered

Other, perhaps more suitable or powerful solutions, are definitely more difficult to implement.

One might be to extend capabilities of multi interceptors (or add another multi special type) to be able to extend themselves with the array of the parent injector.

Lastly, in another train of thought, this problem may delve a little into the "problem" that `ModuleWithProviders` doesn't solve thoroughly, being that provider inclusion is not completely controlled if the module in question imports another module that provides something. And even if the second module has a static `ModuleWithProviders` method for the same argument, you can't (not in a conventional manner) control which module version to import from within these methods. 

i.e. `ModuleWithProviders` can't solve providers in deeper module imports with its intended design.
Hence, the most complicated solution would be to redesign this.

These are a lot less practical solutions, but since there's the section, I might as well mention them.

**Comment by alxhub** [comment]
I would love to see some sort of lazy-loading of interceptors. Thanks for the design proposals! I'm tagging this as a use-case request to allow for discussion.

I feel like for simplicity's sake, rather than having different injectors (created by lazy loading different modules via the router) with different sets of interceptors, a design for this should be based around 2 rules:

1. There is only ever one set of interceptors applied to all requests made by the application.
2. It's possible to _add_ interceptors into this mix on-demand, including ones which are lazily loaded.

**Comment by miguel-leon** [comment]
Hi, thanks for the comment!

I realize now that I didn't specify whether I meant "extending interceptors" as in enlarging the main set of interceptors or as in having different sets of interceptors. I guess modifying the provider in my example a little would allow both cases.

But actually I meant neither (and both?). What I meant was for children module (lazy loaded) injectors to have its own set of interceptors but also benefit from the main ones (a hierarchy).

This was actually the use case in my project. I wanted my child module to apply the application wide interceptors, but also its own "private" set of interceptors. And in retrospective, I don't know if it would have been all OK if this child module would have updated the main set of interceptors. Also the child module was a library (yes, a lazy loaded module in a library with Ivy).

In summary, I think both your set of rules and my use case can be achieved. But I wouldn't want to see this feature restricted to only one way. I rather it to not change.

In any case, both ways have the same (lack of) difficulty to implement, and IMHO, I don't see only one set of interceptors as a superior design, both designs have their pros and cons.

I hope I clarified some things.
Thanks again!
Regards.

**Comment by matheo** [comment]
I just found that our child modules doesn't inherit our root interceptors, 
but have a single `HttpXsrfInterceptor` (I'm not sure where it's coming from, CommonModule, RouterModule, NgRx?)
and it force us to define again all our `HTTP_INTERCEPTORS` which is not ideal because it's not easy to figure out that they were modified by some third-party module and that we need to set them up again.
Why does this happen?
Thanks and best regards!

**Comment by magicmatatjahu** [comment]
Maybe for `HttpClientModule` should export `forChild` macro with config:

```ts
static forChild({ inheritanceOfInterceptors: 'override' | 'concatenate' = 'concatenate' }): ModuleWithProviders<HttpClientModule> {
  const deps = [[new Optional(), HTTP_INTERCEPTORS]];
  if (inheritanceOfInterceptors === 'concatenate') {
    deps.push([new SkipSelf(), new Optional(), HTTP_INTERCEPTORS]);
  }

  return {
    ngModule: HttpClientModule,
    providers: [
      {
        provide: __HTTP_INTERCEPTORS// new private token used only inside HttpModule
        useFactory(interceptors, parentInterceptors) {
          return [...(interceptors || []), ...(parentInterceptors || [])];
        },
        deps,
      }
    ],
  };
}
```

and then change definition of [`HttpInterceptingHandler`](https://github.com/angular/angular/blob/master/packages/common/http/src/module.ts#L31) to: 

```ts
@Injectable()
export class HttpInterceptingHandler implements HttpHandler {
  private chain: HttpHandler|null = null;

  constructor(private backend: HttpBackend, private injector: Injector) {}

  handle(req: HttpRequest<any>): Observable<HttpEvent<any>> {
    if (this.chain === null) {
      const interceptors = this.injector.get(__HTTP_INTERCEPTORS, []); // this line is changed from original code
      this.chain = interceptors.reduceRight(
          (next, interceptor) => new HttpInterceptorHandler(next, interceptor), this.backend);
    }
    return this.chain.handle(req);
  }
}
``` 

It's a solution based @miguel-leon thoughts but showing what an implementation in the `HttpClientModule` itself could look like.

**Comment by miguel-leon** [comment]
Hello, it's been a while since I opened this issue as a feature request.
I've seen some interactions, among them, `HttpHandler` level proposed implementation, and even calling it a bug (https://github.com/angular/angular/issues/39125).
I wanted to say, I still see this as a framework design decision motivated by conceivable use cases.
In that reasoning, I wrote an article commenting on some possibilities regarding that.
Even uploaded a working example to github. So, you're welcome to read.
https://miguel-leon.medium.com/extending-http-interceptors-hierarchically-in-angular-4ca4d25af475

Cheers!

**Comment by destus90** [comment]
We could define root interceptors in `SharedModule` and export `HttpClientModule` from that module. So, in `FeatureModule` we would import `SharedModule` and define its own child interceptors.

Go to https://stackblitz.com/edit/angular-hjzfnu?file=src%2Fapp%2Fshared.module.ts and observe that both headers are applied when you go to `Customers` route. The first interceptor from the `SharedModule` while the second one from the `CustomersModule`.

is it a bad practice to define some interceptors into the `SharedModule`?

**Comment by akyrey** [comment]
Since @miguel-leon solution didn't work for me, I found this article https://indepth.dev/posts/1455/how-to-split-http-interceptors-between-multiple-backends that uses a similar principle and managed to make it work like I wanted.
Sharing here for others getting stuck.

**Comment by vazher21** [comment]
Does not work in angular 15.1.0
getting  following error:

[Error] ERROR – TypeError: interceptor.intercept is not a function. (In 'interceptor.intercept(initialRequest, {
    handle: downstreamRequest => chainTailFn(downstreamRequest, finalHandlerFn)
  })', 'interceptor.intercept' is undefined) — http.mjs:1383
TypeError: interceptor.intercept is not a function. (In 'interceptor.intercept(initialRequest, {
    handle: downstreamRequest => chainTailFn(downstreamRequest, finalHandlerFn)
  })', 'interceptor.intercept' is undefined) — http.mjs:1383(anonymous function) — http.mjs:1383runInContext — core.mjs:7975runInContext — core.mjs:7975doInnerSub — mergeInternals.js:19(anonymous function) — OperatorSubscriber.js:13next — Subscriber.js:31(anonymous function) — innerFrom.js:51_trySubscribe — Observable.js:37(anonymous function) — Observable.js:31errorContext — errorContext.js:19subscribe — Observable.js:22mergeInternals — mergeInternals.js:50(anonymous function) — lift.js:10(anonymous function) — Observable.js:26errorContext — errorContext.js:19subscribe — Observable.js:22(anonymous function) — filter.js:6(anonymous function) — lift.js:10(anonymous function) — Observable.js:26errorContext — errorContext.js:19subscribe — Observable.js:22(anonymous function) — map.js:6(anonymous function) — lift.js:10(anonymous function) — Observable.js:26errorContext — errorContext.js:19subscribe — Observable.js:22getData — shop.component.ts:25:86executeListenerWithErrorHandling — core.mjs:14090wrapListenerIn_markDirtyAndPreventDefault — core.mjs:14123(anonymous function) — platform-browser.mjs:402onInvokeTask — core.mjs:24002onInvokeTask — core.mjs:24300runTask — zone.js:178invokeTask — zone.js:490invokeTask — zone.js:1664globalCallback — zone.js:1695
	handleError (vendor.js:29073)
	next (vendor.js:45205)
	next
	next (vendor.js:5718)
	_next (vendor.js:5687)
	next (vendor.js:5660)
	(anonymous function) (vendor.js:5497)
	errorContext (vendor.js:7937)
	next (vendor.js:5490)
	emit (vendor.js:41197)
	run (polyfills.js:7358)
	onHandleError (vendor.js:44527)
	runTask (polyfills.js:7406)
	invokeTask (polyfills.js:7679)

**Comment by alxhub** [comment]
This issue is resolved by the [`withRequestsMadeViaParent()`](https://angular.io/api/common/http/withRequestsMadeViaParent) feature for `provideHttpClient()`.
